csharp/src/Apache.Arrow/Ipc/ArrowReaderImplementation.cs

```diff
@@ -29,6 +29,13 @@ internal abstract class ArrowReaderImplementation : IDisposable
     {
         public Schema Schema { get; protected set; }
         protected bool HasReadSchema => Schema != null;
+        protected bool HasReadInitialDictionary { get; set; }
+        protected readonly DictionaryMemo _dictionaryMemo;
+
+        public ArrowReaderImplementation()
+        {
+            _dictionaryMemo = new DictionaryMemo();
```

Comment:
I have created the `LazyCreator` class.
`_dictionaryMemo` use it.
However, I am wondering whether this is the best way to lazy load because it need allocate a `LazyCreator` object.
Is there any other good idea?

Reply:
1. I wouldn't use create a new class for this, .NET has https://docs.microsoft.com/dotnet/api/system.lazy-1 already for it.
2. Even better is to not use `Lazy<T>` at all, and just check for null before using it. You can make a private property to make this really easy:

```C#
private DictionaryMemo _dictionaryMemo;

private DictionaryMemo DictionaryMemo => _dictionaryMemo ??= new DictionaryMemo();

// later in code that needs to use DictonaryMemo:

DictionaryMemo.GetDictionaryType(id);
```

This way, nothing new is allocated when we don't see a Dictionary batch in the payload.

Reply:
Thank you, I reimplemented them with the way 2.